dss_metrics/metrics.py

```diff
@@ -0,0 +1,148 @@
+* Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+* Neither the name of Samsung Electronics Co., Ltd. nor the names of its
+  contributors may be used to endorse or promote products derived from this
+  software without specific prior written permission.
+NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
+THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
+NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+from collections import namedtuple
+from prometheus_client import start_http_server, Summary, REGISTRY, Metric
+import argparse
+import json
+import minio_collector
+import multiprocessing
+import nvmftarget_collector
+import time
+import utils
+
+
+MetricInfo = namedtuple("MetricInfo", "key, name, value, tags, timestamp")
+
+
+class MetricsCollector(object):
+    def __init__(self, configs):
+        self.configs = configs
+        self.filter = self.configs['filter']
+        self.whitelist_patterns = utils.get_whitelist_keys()
+
+    def collect(self):
+        metrics = self.get_metrics()
+        for metric in metrics:
+            yield metric
+
+    def get_metrics(self):
+        manager = multiprocessing.Manager()
+        metrics_data_buffer = manager.list()
+        metrics = []
+
+        print("collecting metrics from DSS cluster..")
+
+        num_seconds = self.configs['polling_interval_secs']
+        num_iterations = 1
+
+        try:
+            target_obj = nvmftarget_collector.NVMFTargetCollector(
+                self.configs, num_seconds, num_iterations,
+                self.whitelist_patterns, filter=self.filter
+            )
+        except Exception as error:
+            print(f"Unable to instantiate target collector: {str(error)}")
+
+        try:
+            minio_obj = minio_collector.MinioCollector(
+                self.configs, num_seconds, num_iterations,
+                self.whitelist_patterns, filter=self.filter
+            )
+        except Exception as error:
+            print(f"Unable to instantiate minio collector: {str(error)}")
+
+        target_proc = multiprocessing.Process(
+            target=target_obj.poll_statistics,
```

Comment:
In case of exception, the target_obj is not valid. Similarly for minio_obj as well